### PR TITLE
Modify split config is_keyboard_master/left checks.

### PR DESCRIPTION
**Pull request by daskygit**
## Description

Removes checks from within the main loop, locks left configuration to result on init.

Not fully tested yet but thought I'd share for some feedback.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/split_common/split_util.c [hunk 1]
```diff
@@ -193,12 +170,41 @@ void split_pre_init(void) {
...
+#elif defined(MASTER_RIGHT)
+    split_config.left = !is_keyboard_master();
+#else
+    split_config.left = is_keyboard_master();
 #endif
-    isLeftHand = is_keyboard_left();
+}
+
+__attribute__((weak)) void is_keyboard_master_init(void) {
+    split_config.master = usbIsActive();
```
[the remark this answers is not in the document]
Comment (daskygit) [reply quantum/split_common/split_util.c:182]: The thinking was that it'd ensure setup happened in the correct order if you wanted to handle any custom pin configuration yourself. But you're right that would leave split_config unused at that point.

We could introduce set and get functions for split_config but I don't know if that would be prefered over just making split_config public?

[the remark this answers is not in the document]
Comment (drashna) [reply quantum/split_common/split_util.c:182]: IMO, exposing the functions is the safer option.  If the structs have to change, or there is other behind the scenes stuff, the functions remain the same.  

And we've been moving away from exposing data structures (rgb stuff, for instance)

File: quantum/split_common/split_util.c [hunk 2]
```diff
@@ -131,50 +136,22 @@ static uint8_t peek_matrix_intersection(pin_t out_pin, pin_t in_pin) {
 }
 #endif
 
-__attribute__((weak)) bool is_keyboard_left(void) {
+__attribute__((weak)) void is_keyboard_left_init(void) {
```
Comment (zvecr) [inline quantum/split_common/split_util.c:139]:
```suggestion
__attribute__((weak)) bool is_keyboard_left_impl(void) {
```

File: quantum/split_common/split_util.c [hunk 3]
```diff
@@ -131,50 +136,22 @@ static uint8_t peek_matrix_intersection(pin_t out_pin, pin_t in_pin) {
...
 #if defined(SPLIT_HAND_PIN)
+    setPinInput(SPLIT_HAND_PIN);
+    wait_us(100);
     // Test pin SPLIT_HAND_PIN for High/Low, if low it's right hand
 #    ifdef SPLIT_HAND_PIN_LOW_IS_LEFT
-    return !readPin(SPLIT_HAND_PIN);
+    split_config.left = !readPin(SPLIT_HAND_PIN);
```
Comment (zvecr) [inline quantum/split_common/split_util.c:145]:
```suggestion
    return !readPin(SPLIT_HAND_PIN);
```

File: quantum/split_common/split_util.c [hunk 4]
```diff
@@ -131,50 +136,22 @@ static uint8_t peek_matrix_intersection(pin_t out_pin, pin_t in_pin) {
...
 #    else
-    return readPin(SPLIT_HAND_PIN);
+    split_config.left = readPin(SPLIT_HAND_PIN);
```
Comment (zvecr) [inline quantum/split_common/split_util.c:147]:
```suggestion
    return readPin(SPLIT_HAND_PIN);
```

File: quantum/split_common/split_util.c [hunk 5]
```diff
@@ -131,50 +136,22 @@ static uint8_t peek_matrix_intersection(pin_t out_pin, pin_t in_pin) {
...
 #    endif
 #elif defined(SPLIT_HAND_MATRIX_GRID)
 #    ifdef SPLIT_HAND_MATRIX_GRID_LOW_IS_RIGHT
-    return peek_matrix_intersection(SPLIT_HAND_MATRIX_GRID);
+    split_config.left = peek_matrix_intersection(SPLIT_HAND_MATRIX_GRID);
```
Comment (zvecr) [inline quantum/split_common/split_util.c:151]:
```suggestion
    return peek_matrix_intersection(SPLIT_HAND_MATRIX_GRID);
```

File: quantum/split_common/split_util.c [hunk 6]
```diff
@@ -131,50 +136,22 @@ static uint8_t peek_matrix_intersection(pin_t out_pin, pin_t in_pin) {
...
 #    else
-    return !peek_matrix_intersection(SPLIT_HAND_MATRIX_GRID);
+    split_config.left = !peek_matrix_intersection(SPLIT_HAND_MATRIX_GRID);
```
Comment (zvecr) [inline quantum/split_common/split_util.c:153]:
```suggestion
    return !peek_matrix_intersection(SPLIT_HAND_MATRIX_GRID);
```

File: quantum/split_common/split_util.c [hunk 7]
```diff
@@ -193,12 +170,41 @@ void split_pre_init(void) {
         eeconfig_update_handedness(should_be_left);
     }
 #    endif // defined(INIT_EE_HANDS_LEFT) || defined(INIT_EE_HANDS_RIGHT)
+    split_config.left = eeconfig_read_handedness();
+#elif defined(MASTER_RIGHT)
+    split_config.left = !is_keyboard_master();
+#else
+    split_config.left = is_keyboard_master();
```
Comment (zvecr) [inline quantum/split_common/split_util.c:177]:
```suggestion
    return !is_keyboard_master();
#else
    return is_keyboard_master();
```

File: quantum/split_common/split_util.c [hunk 8]
```diff
@@ -193,12 +170,41 @@ void split_pre_init(void) {
...
+}
+
+__attribute__((weak)) bool is_keyboard_master(void) {
+    return split_config.master;
+}
+
+// this code runs before the keyboard is fully initialized
+void split_pre_init(void) {
+    is_keyboard_master_init();
+    is_keyboard_left_init();
```
Comment (zvecr) [inline quantum/split_common/split_util.c:201]:
```suggestion
    split_config.master = is_keyboard_master_impl();
    split_config.left        = is_keyboard_left_impl();
```

File: quantum/split_common/split_util.c [hunk 9]
```diff
@@ -193,12 +170,41 @@ void split_pre_init(void) {
...
+}
+
+__attribute__((weak)) void is_keyboard_master_init(void) {
+    split_config.master = usbIsActive();
+
+    // Avoid NO_USB_STARTUP_CHECK - Disable USB as the previous checks seem to enable it somehow
+    if (!split_config.master) {
+        usb_disconnect();
+    }
+}
```
Comment (zvecr) [inline quantum/split_common/split_util.c:188]:
```suggestion
__attribute__((weak)) bool is_keyboard_master_impl(void) {
    bool is_master = usbIsActive();

    // Avoid NO_USB_STARTUP_CHECK - Disable USB as the previous checks seem to enable it somehow
    if (!is_master) {
        usb_disconnect();
    }
    return is_master;
}
```

**Review by zvecr — COMMENTED** [review]
I would change it somewhat so that there isnt a need for overrides to have knowledge of the `split_config` variable.

**Review by drashna — APPROVED** [review]

**Review by zvecr — APPROVED** [review]

**Pull request merged**